<?xml version="1.0"?>
<launch>
  <!--
    Use command line to determine potential video settings:
(remove backslashes https://stackoverflow.com/questions/10842131/xml-comments-and)
    v4l2-ctl \-\-list-formats-ext
    v4l2-ctl -d /dev/video0 \-\-list-framesizes=YUYV
  -->
  <arg name="device" default="1" />
  <arg name="width" default="1920" />
  <arg name="height" default="1080" />
  <arg name="small_width" default="480" />
  <arg name="small_height" default="270" />
  <!--
  <arg name="width" default="1280" />
  <arg name="height" default="1024" />
  <arg name="small_width" default="320" />
  <arg name="small_height" default="256" />
  -->
  <!-- The above values need to be width/decimation and height/decimation -->
  <arg name="decimation" default="4" />

  <arg name="frame_rate" default="5" />

  <arg name="use_usb_cam" default="true" />
  <arg name="use_libuvc_camera" default="false" />
  <arg name="use_noise_image" default="false" />

  <group if="$(arg use_usb_cam)">
  <node name="usb_cam" pkg="usb_cam" type="usb_cam_node"
      output="screen" >
    <param name="image_width" value="$(arg width)" />
    <param name="image_height" value="$(arg height)" />
    <param name="framerate" value="$(arg frame_rate)" />
    <param name="video_device" value="/dev/video$(arg device)" />
    <!--param name="pixel_format" value="mjpeg" /-->
    <param name="pixel_format" value="yuyv" />
  </node>
  </group>

  <group if="$(arg use_libuvc_camera)" ns="usb_cam" >
  <node name="uvc_camera" pkg="libuvc_camera" type="camera_node"
      output="screen">
    <!-- Parameters used to find the camera -->
    <param name="vendor" value="0x0"/>
    <param name="product" value="0x0"/>
    <param name="serial" value=""/>
    <!-- If the above parameters aren't unique, choose the first match: -->
    <param name="index" value="$(arg device)"/>

    <!--param name="scanning_mode" value="1" /-->
    <!-- Image size and type -->
    <param name="width" value="$(arg width)"/>
    <param name="height" value="$(arg height)"/>
    <!-- choose whichever uncompressed format the camera supports: -->
    <param name="video_mode" value="uncompressed"/>  <!-- or yuyv/nv12/jpeg -->
    <param name="frame_rate" value="$(arg frame_rate)"/>

    <param name="timestamp_method" value="start"/>  <!-- start of frame -->
    <!--param name="camera_info_url" value="file:///tmp/cam.yaml"/-->

    <param name="auto_exposure" value="0"/>  <!-- use aperture_priority auto exposure -->
    <param name="auto_white_balance" value="false"/>
  </node>
  </group>  <!-- use_libuvc_camera -->

  <group if="$(arg use_noise_image)">
    <node name="noise_image" pkg="vimjay" type="noise_image" >
      <remap from="image" to="image_raw" />
    </node>
  </group>

  <group ns="usb_cam" >

  <!-- http://answers.ros.org/question/255970/how-to-deactivate-image_transport-plugins/ -->
  <group ns="anim" >
    <rosparam param="disable_pub_plugins">
      - 'image_transport/compressed'
      - 'image_transport/compressedDepth'
      - 'image_transport/theora'
    </rosparam>
  </group>

  <group ns="captured_image" >
    <rosparam param="disable_pub_plugins">
      - 'image_transport/compressed'
      - 'image_transport/compressedDepth'
      - 'image_transport/theora'
    </rosparam>
  </group>

  <group ns="diff_image" >
    <rosparam param="disable_pub_plugins">
      - 'image_transport/compressed'
      - 'image_transport/compressedDepth'
      - 'image_transport/theora'
    </rosparam>
  </group>

  <group ns="filtered_image" >
    <rosparam param="disable_pub_plugins">
      - 'image_transport/compressed'
      - 'image_transport/compressedDepth'
      - 'image_transport/theora'
    </rosparam>
  </group>

  <group ns="gray" >
    <rosparam param="disable_pub_plugins">
      - 'image_transport/compressed'
      - 'image_transport/compressedDepth'
      - 'image_transport/theora'
    </rosparam>
  </group>

  <group ns="image_raw" >
    <rosparam param="disable_pub_plugins">
      - 'image_transport/compressed'
      - 'image_transport/compressedDepth'
      - 'image_transport/theora'
    </rosparam>
  </group>

  <group ns="live_image_small" >
    <rosparam param="disable_pub_plugins">
      - 'image_transport/compressed'
      - 'image_transport/compressedDepth'
      - 'image_transport/theora'
    </rosparam>
  </group>

  <group ns="saved_image" >
    <rosparam param="disable_pub_plugins">
      - 'image_transport/compressed'
      - 'image_transport/compressedDepth'
      - 'image_transport/theora'
    </rosparam>
  </group>

  <group ns="saved_image_small" >
    <rosparam param="disable_pub_plugins">
      - 'image_transport/compressed'
      - 'image_transport/compressedDepth'
      - 'image_transport/theora'
    </rosparam>
  </group>

  <!-- capture the input image when triggered -->
  <!--node name="capture_trigger" pkg="vimjay" type="key_trigger" >
    <remap from="trigger" to="single" />
  </node-->

  <!-- TODO(lucasw)
    the save and the deque when triggered may not save the same frame to disk
    and the queue -->
  <node name="save_image" pkg="vimjay" type="save_image"
      output="screen" >
    <remap from="image" to="image_raw" />
    <remap from="single" to="captured_image_trigger" />
  </node>

  <node name="image_deque" pkg="vimjay" type="image_deque"
      output="screen" >
    <remap from="image" to="live_image_small" />
  </node>

  <!-- resize the image down for faster processing -->
  <node name="live_decimate" pkg="nodelet" type="nodelet"
      args="standalone image_proc/crop_decimate" >
    <remap from="camera/image_raw" to="image_raw" />
    <remap from="camera/camera_info" to="camera_info" />
    <param name="decimation_x" value="$(arg decimation)" />
    <param name="decimation_y" value="$(arg decimation)" />
    <remap from="camera_out/image_raw" to="live_image_small" />
    <remap from="camera_out/camera_info" to="camera_info_small" />
  </node>

  <node name="saved_decimate" pkg="nodelet" type="nodelet"
      args="standalone image_proc/crop_decimate" >
    <remap from="camera/image_raw" to="saved_image" />
    <!--remap from="camera/camera_info" to="camera_info" /-->
    <param name="decimation_x" value="$(arg decimation)" />
    <param name="decimation_y" value="$(arg decimation)" />
    <remap from="camera_out/image_raw" to="saved_image_small" />
  </node>

  <node name="blur_image" pkg="vimjay" type="iir_imagev"
      output="screen" >
    <rosparam param="b_coeffs">[0.5, 0.5]</rosparam>
    <remap from="image_0" to="saved_image_small" />
    <remap from="image_1" to="live_image_small" />
  </node>

  <node name="gray" pkg="vimjay" type="color"
      output="screen" >
    <!-- These values need to be width/4 and height/4 -->
    <param name="width" value="$(arg small_width)" />
    <param name="height" value="$(arg small_height)" />
    <remap from="image" to="gray" />
  </node>

  <node name="diff_image" pkg="vimjay" type="iir_imagev"
      output="screen" >
    <rosparam param="b_coeffs">[1.0, 0.5, -0.5]</rosparam>
    <remap from="image_0" to="gray" />
    <remap from="image_1" to="saved_image_small" />
    <remap from="image_2" to="live_image_small" />
    <remap from="filtered_image" to="diff_image" />
  </node>

  <!-- rqt_image_view doesn't support arg yet
  <node name="raw_image_view" pkg="rqt_image_view" type="rqt_image_view"
      args="image_raw" />
  -->
  <!-- TODO(lucasw) re-add hide-title -->
  <!-- lock-perspective and freeze-layout are causing problems
   File "/opt/ros/kinetic/lib/python2.7/dist-packages/rqt_gui/main.py", line 59, in main
    return super(Main, self).main(argv, standalone=standalone, plugin_argument_provider=plugin_argument_provider, plugin_manager_settings_prefix=str(hash(os.environ['ROS_PACKAGE_PATH'])))
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/qt_gui/main.py", line 523, in main
    perspective_manager.import_perspective_from_file(self._options.perspective_file, perspective_manager.HIDDEN_PREFIX + '__cli_perspective_from_file')
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/qt_gui/perspective_manager.py", line 330, in import_perspective_from_file
    self._remove_perspective(perspective_name)
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/qt_gui/perspective_manager.py", line 305, in _remove_perspective
    self._menu_manager.remove_item(name)
AttributeError: 'NoneType' object has no attribute 'remove_ite
  -->
  <node name="rqt_generic" pkg="rqt_gui" type="rqt_gui"
      args="--perspective-file $(find vimjay)/config/stop_motion.perspective"
      output="screen"
      required="false" >
  </node>


  <!--
  <node name="rqt_anim" pkg="rqt_dr_single" type="rqt_dr_single"
      output="screen" >
    <remap from="topic" to="image_deque" />
  </node>
  <node name="rqt_cam" pkg="rqt_dr_single" type="rqt_dr_single"
      output="screen" >
    <remap from="topic" to="v4l2ucp_ddr" />
  </node>
  -->

  <node name="rqt_dr" pkg="rqt_gui" type="rqt_gui"
      args="--perspective-file $(find vimjay)/config/dr.perspective"
      output="screen"
      if="false">
  </node>

  <node name="v4l2ucp" pkg="v4l2ucp" type="v4l2ucp" >
    <param name="device" value="/dev/video$(arg device)" />
  </node>

  </group>
</launch>
